sbt-scalajs-bundler/src/main/scala/scalajsbundler/PackageManager.scala

```diff
@@ -0,0 +1,163 @@
+      case _ =>
+        ()
+    }
+  }
+
+  val packageJsonContents: Map[String, JSON]
+}
+
+trait AddPackagesSupport { this: PackageManager =>
+
+  val addPackagesCommand: String
+  val addPackagesArgs: Seq[String]
+
+  /**
+    * Locally install NPM packages
+    *
+    * @param baseDir The (sub-)project directory which contains yarn.lock
+    * @param installDir The directory in which to install the packages
+    * @param logger sbt logger
+    * @param npmPackages Packages to install (e.g. "webpack", "webpack@2.2.1")
+    */
+  def addPackages(baseDir: File,
+                  installDir: File,
+                  logger: Logger,
+                 )(npmPackages: String*): Unit = {
+    this match {
+      case lfs: LockFileSupport  =>
+        lfs.lockFileRead(baseDir, installDir, logger)
+      case _ =>
+        ()
+    }
+
+    run(addPackagesCommand +: (addPackagesArgs ++ npmPackages): _*)(installDir, logger)
+
+    this match {
+      case lfs: LockFileSupport  =>
+        lfs.lockFileWrite(baseDir, installDir, logger)
+      case _ =>
+        ()
+    }
+  }
+}
+
+trait LockFileSupport {
+  val lockFileName: String
+
+  def lockFileRead(
+    baseDir: File,
+    installDir: File,
+    logger: Logger
+  ): Unit = {
+    val sourceLockFile = baseDir / lockFileName
+    val targetLockFile = installDir / lockFileName
+
+    if (sourceLockFile.exists()) {
+      logger.info("Using lockfile " + sourceLockFile)
+      IO.copyFile(sourceLockFile, targetLockFile)
+    }
+  }
+
+  def lockFileWrite(
+    baseDir: File,
+    installDir: File,
+    logger: Logger
+  ): Unit = {
+    val sourceLockFile = baseDir / lockFileName
+    val targetLockFile = installDir / lockFileName
+
+    if (targetLockFile.exists()) {
+      logger.debug("Wrote lockfile to " + sourceLockFile)
+      IO.copyFile(targetLockFile, sourceLockFile)
+    }
+  }
+}
+
+case class Npm(
+    name: String = "npm",
+    lockFileName: String = "package-lock.json",
+    installCommand: String = "install",
+    installArgs: Seq[String] = Seq.empty,
+    addPackagesCommand: String = "install",
+    addPackagesArgs: Seq[String] = Seq.empty,
+) extends PackageManager
+    with LockFileSupport
+    with AddPackagesSupport {
+  override val packageJsonContents: Map[String, JSON] = Map.empty
+}
+
+case class Yarn(
+    name: String = "yarn",
+    version: Option[String] = None,
+    lockFileName: String = "yarn.lock",
+    installCommand: String = "install",
+    installArgs: Seq[String] = Yarn.DefaultArgs,
+    addPackagesCommand: String = "add",
+    addPackagesArgs: Seq[String] = Yarn.DefaultArgs
+) extends PackageManager
+    with LockFileSupport
+    with AddPackagesSupport {
+  override val packageJsonContents: Map[String, JSON] =
+    version.map(v => Map("packageManager" -> JSON.str(s"$name@$v"))).getOrElse(Map.empty)
+}
+object Yarn {
+  val DefaultArgs: Seq[String] = Seq("--non-interactive", "--mutex", "network")
+}
+
+case class Pnpm(
```

Comment:
Since we are using an OOP approach, this is actually something we don't have to back inside scalajs-bundler! Instead, users who want to use `pnpm` can define something like this in their build. We can document this, and then users won't have to request support for other managers in the future, as they'll be able to define whichever they need themselves. This is similar to the design of `JSEnv`s in Scala.js core.

We'd have to keep `Npm` and `Yarn` in scalajs-bundler for compatibility reasons, but there is no reason to *add* more.

It doesn't mean we can't have a *test* with pnpm. On the contrary, such a test ensures that it is indeed possible to do this in user space.